components/merkledb/src/tests.rs

```diff
@@ -32,3 +33,67 @@ fn should_compile() {
     let _: SparseListIndex<_, ()> = SparseListIndex::new_in_family("", "", &fork);
     let _: ValueSetIndex<_, ()> = ValueSetIndex::new_in_family("", "", &fork);
 }
+
+#[test]
+fn checkpoints() {
+    fn test_checkpoint(db: Box<dyn Database>) {
+        let temp_dir = TempDir::new().unwrap();
+        let root_hashes = {
+            let fork = db.fork();
+            let mut root_hashes = Vec::new();
+            for i in 0..10 {
+                let index_name = format!("idx{}", i);
+                let mut index: ProofListIndex<_, Vec<u8>> = ProofListIndex::new(index_name, &fork);
+                for j in 0..128 {
+                    index.push(vec![0_u8; j]);
+                }
+                root_hashes.push(index.object_hash());
+            }
+            db.merge_sync(fork.into_patch()).unwrap();
+            root_hashes
+        };
+
+        let checkpoint = db.create_checkpoint(&*temp_dir.path().join("db")).unwrap();
+        {
+            let fork = checkpoint.fork();
+            for i in 0..10 {
+                let index_name = format!("idx{}", i);
+                let mut index: ProofListIndex<_, Vec<u8>> = ProofListIndex::new(index_name, &fork);
+                assert_eq!(index.object_hash(), root_hashes[i]);
+                for j in 0..128 {
+                    index.push(vec![0_u8; j]);
+                }
+            }
+            checkpoint.merge_sync(fork.into_patch()).unwrap();
+        }
+
+        {
+            let fork = db.fork();
+            for i in 0..10 {
+                let index_name = format!("idx{}", i);
+                let mut index: ProofListIndex<_, Vec<u8>> = ProofListIndex::new(index_name, &fork);
+                assert_eq!(index.object_hash(), root_hashes[i]);
+                for j in 0..64 {
+                    index.push(vec![0_u8; j]);
+                }
+            }
+            db.merge_sync(fork.into_patch()).unwrap();
+        }
+
+        {
+            let fork = checkpoint.fork();
+            for i in 0..10 {
+                let index_name = format!("idx{}", i);
+                let index: ProofListIndex<_, Vec<u8>> = ProofListIndex::new(index_name, &fork);
+                assert_eq!(index.len(), 256);
+            }
+        }
+    }
+
+    let temp_dir = TempDir::new().unwrap();
+    let rocks_db = RocksDB::open(temp_dir.path().join("db"), &DbOptions::default()).unwrap();
```

Comment:
Just wondering why we need to `join("db")` here, as the temp_dir is different and empty anyway.

Reply:
RocksDb will error in case of existing checkpoint directory. I believe to not to accidentally overwrite something if API is misused.

Reply:
🤔 But you can't overwrite anything because RocksDB will error in case of the same directory. The code is okay, it was just a little surprising to me before I spotted that there are actually two `temp_dir`s.